### PR TITLE
Ebpf 4.0 v22

### PR DESCRIPTION
**Pull request by regit**
Update of #3170 

- [x] I have read the contributing guide lines at https://redmine.openinfosecfoundation.org/projects/suricata/wiki/Contributing
- [x] I have signed the Open Information Security Foundation contribution agreement at https://suricata-ids.org/about/contribution-agreement/
- [x] I have updated the user guide (in doc/userguide/) to reflect the changes made (if applicable)

Link to [redmine](https://redmine.openinfosecfoundation.org/projects/suricata/issues) ticket:

Describe changes:
- Fix CPU redirect map: cpu load balancing was not correctly implemented

[PRScript](https://redmine.openinfosecfoundation.org/projects/suricata/wiki/PRscript) output:
- PR regit: https://buildbot.openinfosecfoundation.org/builders/regit/builds/370
- PR regit-pcap: https://buildbot.openinfosecfoundation.org/builders/regit-pcap/builds/153

### CODE REVIEW

File: doc/userguide/capture-hardware/ebpf-xdp.rst [hunk 1]
```diff
@@ -0,0 +1,318 @@
...
+between kernel and userspace.
+
+eBPF is used for three things in Suricata:
+
+- eBPF filter: any BPF like filter can be developed. An example of filter accepting only packet for some VLANs is provided.
+- eBPF load balancing: provide programmable load balancing. A simple ippair load balancing is provided.
+- XDP programs: suricata can load XDP programs. A bypass program is provided.
+
+Bypass can be implemented in eBPF and XDP. The advantage of XDP is that the packets are dropped at the earliest stage
+possible. So performance are better. But bypassed packets don't reach the network so you can't use this on regular
```
Comment (norg) [inline doc/userguide/capture-hardware/ebpf-xdp.rst]: So performance *is* better?

File: doc/userguide/capture-hardware/ebpf-xdp.rst [hunk 2]
```diff
@@ -0,0 +1,318 @@
...
+ XDP or eXpress Data Path provides a high performance, programmable network data path in the Linux kernel as part of the IO Visor Project. XDP provides bare metal packet processing at the lowest point in the software stack which makes it ideal for speed without compromising programmability. Furthermore, new functions can be implemented dynamically with the integrated fast path without kernel modification.
+
+`More info about XDP <https://www.iovisor.org/technology/xdp>`__
+
+Requirements
+------------
+
+- Newer kernel that supports XDP (4.13.10 in our case)
+- RSS symmetric hashing on the NIC ( Intel 82599ES 10-Gigabit/x520/x540 in our case)
+- In tree kernel drivers NIC drivers
```
Comment (norg) [inline doc/userguide/capture-hardware/ebpf-xdp.rst]: I'm not 100% sure if everyone knows what you mean with that, can you rephrase it?

File: doc/userguide/capture-hardware/ebpf-xdp.rst [hunk 3]
```diff
@@ -0,0 +1,318 @@
...
+
+Prerequisites
+-------------
+
+This guide has been confirmed on Debian/Ubuntu LTS Linux.
```
Comment (norg) [inline doc/userguide/capture-hardware/ebpf-xdp.rst]: maybe add the LTS versions in brackets

File: doc/userguide/capture-hardware/ebpf-xdp.rst [hunk 4]
```diff
@@ -0,0 +1,318 @@
...
+~~~~~
+
+Make sure you have clang (>3.9) installed on the system  ::
+
+ apt-get install clang
+
+BPF
+~~~
+
+Suricata use libbpf to interact with eBPF and XDP. This library is available
```
Comment (norg) [inline doc/userguide/capture-hardware/ebpf-xdp.rst]: *uses*

File: doc/userguide/capture-hardware/ebpf-xdp.rst [hunk 5]
```diff
@@ -0,0 +1,318 @@
...
+
+If you plan to use eBPF or XDP for a kernel/hardware level bypass, you need to do
+the following:
+
+First, enable `bypass` in the `stream` section ::
+
+ stream:
+   bypass: true
+
+If you with you can also bypass encrypted flow by setting no-reassemble to yes
```
Comment (norg) [inline doc/userguide/capture-hardware/ebpf-xdp.rst]: If you *want* instead of *with* I guess?

File: doc/userguide/capture-hardware/ebpf-xdp.rst [hunk 6]
```diff
@@ -0,0 +1,318 @@
...
+from one card to the second card without going by the kernel network stack.
+
+Setup symmetric hashing on the NIC
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Follow these instructions closely for desired result::
+
+ ifconfig eth3 down
+
+Use in tree kernel drivers, XDP support is not available in out of source Intel drivers.
```
Comment (norg) [inline doc/userguide/capture-hardware/ebpf-xdp.rst]: *out of source* drivers? You mean those not within the kerneltree?

File: doc/userguide/capture-hardware/ebpf-xdp.rst [hunk 7]
```diff
@@ -0,0 +1,318 @@
...
+ ethtool -L eth3 combined 16
+ ethtool -K eth3 rxhash on 
+ ethtool -K eth3 ntuple on
+ ifconfig eth3 up
+ ./set_irq_affinity 0-15 eth3
+ ethtool -X eth3 hkey 6D:5A:6D:5A:6D:5A:6D:5A:6D:5A:6D:5A:6D:5A:6D:5A:6D:5A:6D:5A:6D:5A:6D:5A:6D:5A:6D:5A:6D:5A:6D:5A:6D:5A:6D:5A:6D:5A:6D:5A equal 16
+ ethtool -x eth3
+ ethtool -n eth3
+
+In the above set up you are free to use any recent ``set_irq_affinity`` script. It is available in any Intel x520/710 NIC sources driver download.
```
Comment (norg) [inline doc/userguide/capture-hardware/ebpf-xdp.rst]: *setup* instead of *set up*?

File: doc/userguide/capture-hardware/ebpf-xdp.rst [hunk 8]
```diff
@@ -0,0 +1,318 @@
...
+
+**NOTE:**
+We use a special low entropy key for the symmetric hashing. `More info about the research for symmetric hashing set up <http://www.ndsl.kaist.edu/~kyoungsoo/papers/TR-symRSS.pdf>`_
+
+Disable an NIC offloading
```
Comment (norg) [inline doc/userguide/capture-hardware/ebpf-xdp.rst]: *any* instead of *an*?

File: doc/userguide/capture-hardware/ebpf-xdp.rst [hunk 9]
```diff
@@ -0,0 +1,318 @@
...
+~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Run the following to disable offloading ::
```
Comment (norg) [inline doc/userguide/capture-hardware/ebpf-xdp.rst]: +command

File: doc/userguide/capture-hardware/ebpf-xdp.rst (hunk 6 above)
Comment (regit) [reply doc/userguide/capture-hardware/ebpf-xdp.rst]: Fixed: Intel website driver don't have the feature

**Comment by regit** [comment]
@norg Thanks for the documentation review!

**Comment by regit** [comment]
Closing in favor of #3180
